### PR TITLE
Add guided decoding to TGIS gRPC API

### PR DESCRIPTION
**Pull request by njhill**
Within the existing `decoding` request parameter section:
  
  ```protobuf
  enum ResponseFormat {
    // Plain text, no constraints
    TEXT = 0;
    // Valid json
    JSON = 1;
  }

  message StringChoices {
    repeated string choices = 1;
  }

  // Mutually-exclusive guided decoding options
  oneof guided {
    // Output will be in the specified format
    ResponseFormat format = 3;
    // Output will follow the provided JSON schema
    string json_schema = 4;
    // Output will follow the provided regex pattern
    string regex = 5;
    // Output will be exactly one of the specified choices
    StringChoices choice = 6;
    // Output will follow the provided context free grammar
    string grammar = 7;
  }
```

### CODE REVIEW

File: proto/generation.proto [hunk 1]
```diff
@@ -99,6 +99,31 @@ message DecodingParameters {
...
+  // Mutually-exclusive guided decoding options
+  oneof guided {
+    // Output will be in the specified format
+    ResponseFormat format = 3;
+    // Output will follow the provided JSON schema
+    string json_schema = 4;
+    // Output will follow the provided regex pattern
+    string regex = 5;
+    // Output will be exactly one of the specified choices
+    StringChoices choice = 6;
```
Comment (njhill) [inline proto/generation.proto:123]: Unfortunately you cannot have `repeated` fields directly within `oneof`s :(

https://github.com/protocolbuffers/protobuf/issues/2592#issuecomment-293978001

File: vllm/entrypoints/grpc/grpc_server.py [hunk 2]
```diff
@@ -118,7 +120,8 @@ def __init__(self, engine: AsyncLLMEngine, args: argparse.Namespace):
 
     async def _post_init(self):
         self.config = await self.engine.get_model_config()
-        self.tokenizer_group = await self.engine.get_tokenizer_group()
+        # self.tokenizer_group = await self.engine.get_tokenizer_group()
+        self.tokenizer_group = self.engine.engine.tokenizer
```
Comment (maxdebayser) [inline vllm/entrypoints/grpc/grpc_server.py:124]: I've seen versions of the code where the get_tokenizer_group function exists and others where it doesn't. What's happening with this function?

**Review by maxdebayser — APPROVED** [review]
Since the bug reported in issue https://github.ibm.com/ai-foundation/fmaas-inference-server/issues/718 is not cause by the code in this PR, I think we can merge it and fix the problem in another PR.

File: vllm/tgis_utils/guided_decoding.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    We cache logit processors by (guide, tokenizer), and on cache hit
+    we make a shallow copy to reuse the same underlying FSM.
+    """
+    guide, mode = _get_guide_and_mode(decoding_params)
+    if not guide:
+        return None
+
+    if outlines_decoding.global_thread_pool is None:
+        outlines_decoding.global_thread_pool = (
+            concurrent.futures.ThreadPoolExecutor(max_workers=2))
```
[the remark this answers is not in the document]
Comment (maxdebayser) [reply vllm/tgis_utils/guided_decoding.py:30]: It's the same code as here: https://github.com/IBM/vllm/blob/4af59d38c5bd79ed4f89e226af0ecb59a6bbe7d7/vllm/model_executor/guided_decoding/outlines_decoding.py#L71 . If I'm not mistaken, only the construction of the logits processor happens in another thread. But if the logits processor is cached, I'm not sure what's the benefit of having another thread build the object.

File: vllm/entrypoints/grpc/grpc_server.py (hunk 2 above)
Comment (joerunde) [reply vllm/entrypoints/grpc/grpc_server.py:124]: @maxdebayser that's from this upstream PR https://github.com/vllm-project/vllm/pull/3512

It didn't get merged in a timely manner and is now buried in conflicts :(

File: vllm/tgis_utils/guided_decoding.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/tgis_utils/guided_decoding.py:30]: Yes that's right. The code is just the same as that in the http API. It's dispatched to a threadpool to avoid blocking the asyncio event loop, but I think it could be made more efficient since we only care about this in the case that the LP is not already cached. In any case we can fix that as a follow-on since we need to fix that related concurrency bug anyhow.
